ribbon.py

```diff
@@ -0,0 +1,545 @@
+        self.hsv_upper = (0, 0, 0)
+        self.DRIVING = True
+        self.tracking = False
+        self.finished = False
+        self.i = 0
+        self.start()
+
+    def run(self):
+        # This method runs in a separate thread
+        while not self.terminated:
+            # Wait for an image to be written to the stream
+            if self.event.wait(1):
+                try:
+                    # Read the image and do some processing on it
+                    self.stream.seek(0)
+                    self.process_image(self.stream.array, self.screen)
+                finally:
+                    # Reset the stream and event
+                    self.stream.seek(0)
+                    self.stream.truncate()
+                    self.event.clear()
+
+    def display_marker(self, ribbon_image, marker_contour):
+        cropped_ribbon_image = ribbon_image[0:self.MARKER_CROP_HEIGHT, (self.image_centre_x - self.MARKER_CROP_WIDTH/2):(self.image_centre_x + self.MARKER_CROP_WIDTH/2)]
+        cropped_ribbon_image = cv2.cvtColor(cropped_ribbon_image, cv2.COLOR_GRAY2RGB)
+        cropped_ribbon_image = cv2.cvtColor(cropped_ribbon_image, cv2.COLOR_RGB2HSV)
+        marker_min_colour, marker_max_colour = self.colour_bounds.get(self.MARKER_COLOUR)
+        cv2.drawContours(cropped_ribbon_image, [marker_contour], -1, marker_max_colour, -1)
+        cropped_ribbon_image = cv2.cvtColor(cropped_ribbon_image, cv2.COLOR_HSV2BGR)
+        frame = pygame.surfarray.make_surface(cv2.flip(cropped_ribbon_image, 1))
+        screen = pygame.display.get_surface()
+        image_offset = (320-self.MARKER_CROP_WIDTH)/2
+        screen.blit(frame, (self.CROP_HEIGHT, image_offset))
+        pygame.display.update()
+
+    def check_for_aruco(self, image):
+        gray = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
+        parameters =  aruco.DetectorParameters_create()
+        #lists of ids and the corners beloning to each id
+        corners, ids, rejectedImgPoints = aruco.detectMarkers(gray, self.small_dict, parameters=parameters)
+        if ids != None:
+            if len(ids)>1:
+                logger.info( "found %d aruco markers" % len(ids))
+                closest_marker_index = 0
+                closest_marker_width = 0
+                for marker_number in range(0, len(ids)):
+                    #more than one aruco marker found. cycle through them all, look for biggest
+                    m = marker_number #to keep next few lines short
+                    width = math.sqrt(math.pow(corners[m][0][0][0]-corners[m][0][1][0],2)
+                        +math.pow(corners[m][0][0][1]-corners[m][0][1][1],2))
+                    if width > closest_marker_width:
+                        closest_marker_width = width
+                        closest_marker_index = marker_number
+                        closest_marker_x = sum([arr[0] for arr in corners[m][0]])  / 4
+                        closest_marker_y = sum([arr[1] for arr in corners[m][0]])  / 4
```

Comment:
I'm sure we could avoid this double iteration, but I'm not going to go there now.